[DO NOT MERGE] BYOC on AWS+GCP draft documentation (#37021)
MaterializeInc/materialize: doc/user/content/byoc/byoc-aws-draft.md

- **Compliance**: SOC 2 / HIPAA controls that require infrastructure ownership.
- **Cost visibility**: compute and storage are billed directly to your AWS account.

{{< note >}}
BYOC is set up together with the Materialize team. This guide describes the
steps; your Materialize contact coordinates the handoff. BYOC on AWS requires an
active BYOC subscription.
{{< /note >}}

## How it works

Your environment runs entirely in your AWS account. Materialize provisions and
operates it across your account boundary using an IAM role you create, scoped by
a permission boundary you control. Only operational telemetry (logs and metrics,
with sensitive values redacted) leaves your account so Materialize can monitor
and support the deployment.

## Prerequisites

- An AWS account, and the AWS region you want to run in.
- Permission to run CloudFormation and create IAM roles in that account.
- An active Materialize BYOC subscription.

## Step 1: Share setup details

Materialize sends you a **pre-filled CloudFormation quick-create link**. Opening
it loads the CloudFormation console in your own AWS account with the required

> values (your Materialize account ID, an external ID, and a deployer role ARN)

Reviewer: ```suggestion
values (Materialize's AWS account ID, an external ID, a deployer role ARN, and a name prefix)
```

already populated, so there is nothing to copy by hand. You confirm your target
region, VPC CIDR, and availability zones with your Materialize contact for
provisioning.

## Step 2: Launch the stack

Open the quick-create link while signed in to the AWS account and region where
you want Materialize to run. The CloudFormation form is already populated, so you
review it and launch the stack (about 5 to 10 minutes). It creates exactly two
objects:

> - **An IAM role** (`MaterializeBYOC`) that Materialize assumes to manage your

Reviewer: ```suggestion
- **An IAM role** (`MaterializeBYOCRole`) that Materialize assumes to manage your
```

  environment.

> - **A permission boundary** (`MaterializeBYOCBoundary`) that caps what that role

Reviewer: To be clear, these are the names in the template, not the names they will have in AWS. The names in AWS will be `${NamePrefix}role` and `${NamePrefix}boundary`

Author: Removed the hardcoded names since the real ones are prefix-based and the prefix isn't 'Materialize

Establish an AWS PrivateLink connection from your VPC to reach the environment
privately.

## Security model

- **No standing credentials.** Materialize assumes the role you create using
  short-lived STS credentials with an external ID for confused-deputy
  protection. No long-lived keys are stored.
- **A permission ceiling.** A transitive permission boundary prevents Materialize
  from acting beyond the agreed scope, creating roles outside it, or removing its
  own boundary.
- **Single-tenant isolation.** Dedicated EKS, VPC, RDS, and S3 in your account.
  Data at rest is encrypted with your own KMS keys.
- **You hold the kill switch.** Revoke the role's trust policy at any time to cut
  access. While access is revoked, your environment keeps serving queries but
  cannot be upgraded, scaled, or repaired until access is restored.
- **Audited support access.** Materialize support uses scoped, time-bound,
  audited access. Every action the role takes is recorded in your own AWS
  CloudTrail.

## Observability

Your data stays in your account. Operational telemetry is collected so
Materialize can monitor and support your deployment, and sensitive values in logs

> are redacted before anything leaves your account. A copy of your metrics and logs
> is also kept in your account (CloudWatch for logs and a Prometheus-compatible
> store for metrics) so you can query it with your own tools.

Reviewer: ```suggestion
are redacted before anything leaves your account. A copy of your metrics and logs
is also kept in your account (Loki for logs and a Prometheus-compatible
store for metrics) so you can query it with your own tools.
```

I think the plan is now to ship Loki. CC @Alphadelta14 to confirm.

Reviewer 2: Loki is the planned default, with SM support in July. Other log providers may be desired by request / as demanded (CloudWatch Logs is AWS native but dang it's expensive)
